src/hotspot/share/oops/oop.cpp

```diff
@@ -220,12 +221,12 @@ jdouble oopDesc::double_field_acquire(int offset) const               { return A
 void oopDesc::release_double_field_put(int offset, jdouble value)     { Atomic::release_store(field_addr<jdouble>(offset), value); }
 
 #ifdef ASSERT
-bool oopDesc::size_might_change() {
+bool oopDesc::size_might_change(Klass* klass) {
   // UseParallelGC and UseG1GC can change the length field
   // of an "old copy" of an object array in the young gen so it indicates
   // the grey portion of an already copied array. This will cause the first
   // disjunct below to fail if the two comparands are computed across such
   // a concurrent change.
-  return Universe::heap()->is_stw_gc_active() && is_objArray() && is_forwarded() && (UseParallelGC || UseG1GC);
+  return Universe::heap()->is_stw_gc_active() && klass->is_objArray_klass() && is_forwarded() && (UseParallelGC || UseG1GC);
```

Comment:
Is this still true after the recent changes like JDK-8311163? It might be worth waiting for.

Reply:
Which ones are remaining? JDK-8337709 implemented the same change for G1 GC before JDK-8311163.

The full collectors/g1 marking do not modify the length fields but have multiple separate queues which is a different issue.

If I am correct, this whole method is unnecessary now.

Reply:
If you've already fixed this for GC then I agree that we could remove this.

Reply:
This seems like something that should be done as a separate patch that gets pushed before this PR.

Reply:
Will remove in JDK-8340119.